### ISSUE TITLE
Linq2db fails with "The method or operation is not implemented."

### ISSUE THREAD

**Issue by v-zubritsky** [linq2db/linq2db#2895]
Hi! I encounter `NotImplementedException` from [here](https://github.com/linq2db/linq2db/blob/d9832b6e8dd60e4946e26c07f5e4ac241d167f11/Source/LinqToDB/Linq/Builder/SelectContext.cs#L876) I believe. 

Model is probably not in the simplest form to reproduce the error, but I tried to remove all the unnecessary details.

`The method or operation is not implemented.`
```
   at LinqToDB.Linq.Builder.SelectContext.GetContext(Expression expression, Int32 level, BuildInfo buildInfo)
   at LinqToDB.Linq.Builder.PassThroughContext.GetContext(Expression expression, Int32 level, BuildInfo buildInfo)
   at LinqToDB.Linq.Builder.SelectContext.<>c__DisplayClass55_0.<GetContext>b__0(IBuildContext ctx, Expression ex, Int32 l)
   at LinqToDB.Linq.Builder.SelectContext.ProcessScalar[T](Expression expression, Int32 level, Func`4 action, Func`1 defaultAction, Boolean throwOnError)
   at LinqToDB.Linq.Builder.SelectContext.GetContext(Expression expression, Int32 level, BuildInfo buildInfo)
   at LinqToDB.Linq.Builder.SelectManyBuilder.SelectManyContext.GetContext(Expression expression, Int32 level, BuildInfo buildInfo)
   at LinqToDB.Linq.Builder.TableBuilder.BuildSequence(ExpressionBuilder builder, BuildInfo buildInfo)
   at LinqToDB.Linq.Builder.ExpressionBuilder.BuildSequence(BuildInfo buildInfo)
   at LinqToDB.Linq.Builder.SelectManyBuilder.BuildMethodCall(ExpressionBuilder builder, MethodCallExpression methodCall, BuildInfo buildInfo)
   at LinqToDB.Linq.Builder.MethodCallBuilder.BuildSequence(ExpressionBuilder builder, BuildInfo buildInfo)
   at LinqToDB.Linq.Builder.ExpressionBuilder.BuildSequence(BuildInfo buildInfo)
   at LinqToDB.Linq.Builder.SelectManyBuilder.BuildMethodCall(ExpressionBuilder builder, MethodCallExpression methodCall, BuildInfo buildInfo)
   at LinqToDB.Linq.Builder.MethodCallBuilder.BuildSequence(ExpressionBuilder builder, BuildInfo buildInfo)
   at LinqToDB.Linq.Builder.ExpressionBuilder.BuildSequence(BuildInfo buildInfo)
   at LinqToDB.Linq.Builder.ExpressionBuilder.Build[T]()
   at LinqToDB.Linq.Query`1.CreateQuery(IDataContext dataContext, Expression expr)
   at LinqToDB.Linq.Query`1.GetQuery(IDataContext dataContext, Expression& expr)
   at LinqToDB.Linq.ExpressionQuery`1.GetQuery(Expression& expression, Boolean cache)
   at LinqToDB.Linq.ExpressionQuery`1.System.Collections.Generic.IEnumerable<T>.GetEnumerator()
   at System.Collections.Generic.List`1..ctor(IEnumerable`1 collection)
   at System.Linq.Enumerable.ToList[TSource](IEnumerable`1 source)
   at LinqToDB.Linq.Builder.EagerLoading.<>c__DisplayClass41_0`2.<RegisterPreambles>b__0(IDataContext dc, Expression expr, Object[] ps)
   at LinqToDB.Linq.Builder.ExpressionBuilder.<>c__DisplayClass187_0`1.<RegisterPreamble>b__0(IDataContext dc, Expression e, Object[] ps)
   at LinqToDB.Linq.Query.InitPreambles(IDataContext dc, Expression rootExpression, Object[] ps)
   at LinqToDB.Linq.ExpressionQuery`1.System.Collections.Generic.IEnumerable<T>.GetEnumerator()
   at System.Linq.Buffer`1..ctor(IEnumerable`1 source)
   at System.Linq.Enumerable.ToArray[TSource](IEnumerable`1 source)
   at Repro.Program.Main(String[] args) in D:\dev\Repro\Program.cs:line 22
```

### Steps to reproduce
Here is the problematic query
```c#
var connectionString = "Data Source=.;Initial Catalog=Repro;Integrated Security=True; MultipleActiveResultSets=True";

var builder = new LinqToDbConnectionOptionsBuilder();
builder.UseSqlServer(connectionString);

var dbContext = new DbContext(builder.Build());

var failing = dbContext.Requests
	.Select(r => new
	{
		DocumentName = r.User.Admin.EmailAdminAssociations
			.Select(ea => ea.Email.InternalEmail)
			.Select(ie => new
			{
				Names = ie.Email.EmailAttachmentAssociations
					.SelectMany(eba => eba.Attachment.Documents.Select(aa => aa.Name)),
			})
			.FirstOrDefault(),
	})
	.ToArray();
```

Data connection type looks this way
```c#
public class DbContext : DataConnection
{
	public DbContext(LinqToDbConnectionOptions options) : base(options: options) { }

	public ITable<Request> Requests => GetTable<Request>();
	public ITable<InternalEmail> InternalEmails => GetTable<InternalEmail>();
	public ITable<User> Users => GetTable<User>();
	public ITable<Admin> Admins => GetTable<Admin>();
	public ITable<EmailAdminAssociation> EmailAdminAssociations => GetTable<EmailAdminAssociation>();
	public ITable<Email> Emails => GetTable<Email>();
	public ITable<EmailAttachmentAssociation> EmailAttachmentAssociations => GetTable<EmailAttachmentAssociation>();
	public ITable<Attachment> Attachments => GetTable<Attachment>();
	public ITable<Document> Documents => GetTable<Document>();
}
```

And here is the model
```c#
[Table]
public class InternalEmail 
{
	[PrimaryKey, Column]
	public int Id { get; set; }

	[Column]
	public int? RequestId { get; set; }

	[Column]
	public int? UserId { get; set; }

	[Association(ThisKey = nameof(RequestId), OtherKey = nameof(Repro.Request.Id))]
	public virtual Request Request { get; set; }

	[Association(ThisKey = nameof(UserId), OtherKey = nameof(Repro.User.Id))]
	public virtual User User { get; set; }

	[Association(ThisKey = nameof(Id), OtherKey = nameof(Repro.Email.Id))]
	public virtual Email Email { get; set; }
}

[Table]
public class Request
{
	[PrimaryKey, Identity, Column]
	public int Id { get; set; }

	[Column]
	public int UserId { get; set; }

	[Association(ThisKey = nameof(UserId), OtherKey = nameof(Repro.User.Id))]
	public virtual User User { get; set; }

	[Association(ThisKey = nameof(Id), OtherKey = nameof(InternalEmail.Id))]
	public virtual ICollection<InternalEmail> InternalEmails { get; set; }
}

[Table]
public class User
{
	[PrimaryKey, Identity, Column]
	public int Id { get; set; }

	[Association(ThisKey = nameof(Id), OtherKey = nameof(Repro.Admin.Id))]
	public virtual Admin Admin { get; set; }
}

[Table]
public class Admin
{
	[PrimaryKey, Column]
	public int Id { get; set; }

	[Association(ThisKey = nameof(Id), OtherKey = nameof(EmailAdminAssociation.AdminId))]
	public virtual ICollection<EmailAdminAssociation> EmailAdminAssociations { get; set; }
}

[Table]
public class EmailAdminAssociation
{
	[PrimaryKey(1)]
	[Column]
	public int EmailId { get; set; }

	[PrimaryKey(2)]
	[Column]
	public int AdminId { get; set; }

	[Association(ThisKey = nameof(EmailId), OtherKey = nameof(Repro.Email.Id))]
	public virtual Email Email { get; set; }

	[Association(ThisKey = nameof(AdminId), OtherKey = nameof(Repro.Admin.Id))]
	public virtual Admin Admin { get; set; }
}

[Table]
public class Email
{
	[PrimaryKey, Identity, Column]
	public int Id { get; set; }

	[Association(ThisKey = nameof(Id), OtherKey = nameof(Repro.InternalEmail.Id))]
	public virtual InternalEmail InternalEmail { get; set; }

	[Association(ThisKey = nameof(Id), OtherKey = nameof(EmailAttachmentAssociation.EmailId))]
	public virtual ICollection<EmailAttachmentAssociation> EmailAttachmentAssociations { get; set; }
}

[Table]
public class EmailAttachmentAssociation
{
	[PrimaryKey(1)]
	[Column]
	public int EmailId { get; set; }

	[PrimaryKey(2)]
	[Column]
	public int AttachmentId { get; set; }

	[Association(ThisKey = nameof(AttachmentId), OtherKey = nameof(Repro.Attachment.Id))]
	public virtual Attachment Attachment { get; set; }

	[Association(ThisKey = nameof(EmailId), OtherKey = nameof(Repro.Email.Id))]
	public virtual Email Email { get; set; }
}

[Table]
public class Attachment
{
	[PrimaryKey, Identity, Column]
	public int Id { get; set; }

	[Association(ThisKey = nameof(Id), OtherKey = nameof(Document.AttachmentId))]
	public virtual ICollection<Document> Documents { get; set; }
}

[Table]
public class Document
{
	[PrimaryKey(1)]
	[Column]
	public int AttachmentId { get; set; }

	[PrimaryKey(2)]
	[Column]
	public int Position { get; set; }

	[Column]
	public string Name { get; set; }
}
```

Whole solution to reproduce the issue is here
[Repro.zip](https://github.com/linq2db/linq2db/files/6149213/Repro.zip)

### Environment details
linq2db version: 3.2.3
Database Server: MS SQL Server 14.0.1000
Database Provider: Default Sql Server provider, which is "SqlServer", I believe
Operating system: Windows 10.0.18363 
.NET Framework: 4.8

**Comment by sdanyliv** [comment]
Thanks for reporting. Will check.

**Comment by v-zubritsky** [comment]
@MaceWindu I noticed, that you added this issue to the 3.3.0 milestone. Is the root cause of it known?

**Comment by sdanyliv** [comment]
@v-zubritsky, actually I have fix, but working on improvement of such case.

**Comment by v-zubritsky** [comment]
@sdanyliv That's great news, thank you.

**Comment by v-zubritsky** [comment]
@sdanyliv Fix was postponed and will be available with 3.4.0 release, wasn't it? Is the 3.4.0 expected date known yet?

Hoped to get it fixed this time and need to understand now whether It's ok to wait a bit more or I have to rewrite this query in some other way, as this one and a few other expressions, I also created issues for, block me from migration from EF to EF Core + Linq2db.
Thanks. Appreciate you work.

**Comment by MaceWindu** [comment]
We had no time to fix it for upcoming release, so it will be fixed in next one. Wether it will be 3.4 or 3.3.1

**Comment by v-zubritsky** [comment]
No pressure, I just would like to know what is the expected time frame.

**Comment by MaceWindu** [comment]
We have two features almost ready for release so probably we will have another one in 1-2 weeks. But this issue is quite complex so it mostly depends on how much time it will take to fix it.

**Comment by v-zubritsky** [comment]
Got it, thanks a lot.

And now I'm wondering whether you folks work full-time on it, the fastest response time I've ever seen :)

**Comment by MaceWindu** [comment]
Release day... Have to monitor it :)

**Comment by MaceWindu** [comment]
Reopen as we need to introduce better fix, that doesn't break other queries (see #3384)
